### PR TITLE
ENH Add generic types

### PR DESCRIPTION
**Pull request by GuySartorelli**
Requires https://github.com/silverstripe/silverstripe-framework/pull/11108 for these generics to take effect.

## Issue
- https://github.com/silverstripe/silverstripe-framework/issues/11066

### CODE REVIEW

File: src/Versioned.php [hunk 1]
```diff
@@ -36,8 +36,9 @@
  * Note: This extension relies on the object also having the {@see Ownership} extension applied.
  *
  * @property int $Version
- * @property DataObject|RecursivePublishable|Versioned $owner
  * @mixin RecursivePublishable
+ *
+ * @extends DataExtension<DataObject&RecursivePublishable&Versioned>
```
Comment (emteknetnz) [inline src/Versioned.php]:
```suggestion
 * @extends DataExtension<DataObject&RecursivePublishable&static>
```

File: src/Versioned.php [hunk 2]
```diff
@@ -2363,13 +2360,13 @@ public static function reading_archived_date($date, $stage = self::DRAFT)
...
-     * @param string $class The name of the class.
+     * @template T of DataObject
+     * @param class-string<T> $class The name of the class.
      * @param string $stage The name of the stage.
      * @param string $filter A filter to be inserted into the WHERE clause.
      * @param boolean $cache Use caching.
      * @param string $sort A sort expression to be inserted into the ORDER BY clause.
-     *
-     * @return DataObject
+     * @return T&Versioned
```
Comment (emteknetnz) [inline src/Versioned.php]:
```suggestion
     * @return T&static
```

File: src/Versioned.php [hunk 3]
```diff
@@ -2657,9 +2654,10 @@ public function rollbackSingle($version)
     /**
      * Return the latest version of the given record.
      *
-     * @param string $class
+     * @template T of DataObject
+     * @param class-string<T> $class
      * @param int $id
-     * @return DataObject
+     * @return T&Versioned
```
Comment (emteknetnz) [inline src/Versioned.php]:
```suggestion
     * @return T&static
```

File: src/VersionedStateExtension.php [hunk 4]
```diff
@@ -11,7 +11,8 @@
 /**
  * Persists versioned state between requests via querystring arguments
  *
- * @property RequestHandler|DataObject $owner
+ * @template T of RequestHandler|DataObject
+ * @extends Extension<T>
```
Comment (emteknetnz) [inline src/VersionedStateExtension.php:15]:
```suggestion
 * @extends Extension<RequestHandler|DataObject>
```
Wouldn't this do the same thing?

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/VersionedStateExtension.php (hunk 4 above)
Comment (ChloeSartorelli) [reply src/VersionedStateExtension.php:15]: Nope. Having the `@template T` there means subclasses can say "Use my type instead" using `@extends`.
`VersionedTestSessionExtension` is a subclass of `VersionedStateExtension` and has its own type, so `VersionedStateExtension` has to use `@template T` to allow that to happen.

File: src/Versioned.php (hunk 3 above)
Comment (ChloeSartorelli) [reply src/Versioned.php]: Done

File: src/Versioned.php (hunk 2 above)
Comment (ChloeSartorelli) [reply src/Versioned.php]: Done

File: src/Versioned.php (hunk 1 above)
Comment (ChloeSartorelli) [reply src/Versioned.php]: Done

**Review by emteknetnz — APPROVED** [review]

**Pull request merged**